go/vt/topo/test/trylock.go

```diff
@@ -0,0 +1,182 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"context"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"vitess.io/vitess/go/vt/topo"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+)
+
+// checkLock checks we can lock / unlock as expected. It's using a keyspace
+// as the lock target.
+func checkTryLock(t *testing.T, ts *topo.Server) {
+	ctx := context.Background()
+	if err := ts.CreateKeyspace(ctx, "test_keyspace", &topodatapb.Keyspace{}); err != nil {
+		require.Fail(t, "CreateKeyspace fail", err.Error())
+	}
+
+	conn, err := ts.ConnForCell(context.Background(), topo.GlobalCell)
+	if err != nil {
+		require.Fail(t, "ConnForCell(global) failed", err.Error())
+	}
+
+	t.Log("===      checkTryLockTimeout")
+	checkTryLockTimeout(ctx, t, conn)
+
+	t.Log("===      checkTryLockMissing")
+	checkTryLockMissing(ctx, t, conn)
+
+	t.Log("===      checkTryLockUnblocks")
+	checkTryLockUnblocks(ctx, t, conn)
+}
+
+// checkTryLockTimeout test the non-blocking nature of TryLock
+func checkTryLockTimeout(ctx context.Context, t *testing.T, conn topo.Conn) {
+	keyspacePath := path.Join(topo.KeyspacesPath, "test_keyspace")
+	lockDescriptor, err := conn.TryLock(ctx, keyspacePath, "")
+	if err != nil {
+		require.Fail(t, "TryLock failed", err.Error())
+	}
+
+	// We have the lock, list the keyspace directory.
+	// It should not contain anything, except Ephemeral files.
+	entries, err := conn.ListDir(ctx, keyspacePath, true /*full*/)
+	if err != nil {
+		require.Fail(t, "ListDir failed: %v", err.Error())
+	}
+	for _, e := range entries {
+		if e.Name == "Keyspace" {
+			continue
+		}
+		if e.Ephemeral {
+			t.Logf("skipping ephemeral node %v in %v", e, keyspacePath)
+			continue
+		}
+		// Non-ephemeral entries better have only ephemeral children.
+		p := path.Join(keyspacePath, e.Name)
+		entries, err := conn.ListDir(ctx, p, true /*full*/)
+		if err != nil {
+			require.Fail(t, "ListDir failed", err.Error())
+		}
+		for _, e := range entries {
+			if e.Ephemeral {
+				t.Logf("skipping ephemeral node %v in %v", e, p)
+			} else {
+				require.Fail(t, "non-ephemeral DirEntry")
+			}
+		}
+	}
+
+	// test we can't take the lock again
+	fastCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	if _, err := conn.TryLock(fastCtx, keyspacePath, "again"); !topo.IsErrType(err, topo.NodeExists) {
+		require.Fail(t, "TryLock failed", err.Error())
+	}
+	cancel()
+
+	// test we can interrupt taking the lock
+	interruptCtx, cancel := context.WithCancel(ctx)
+	go func() {
+		time.Sleep(1 * time.Second)
+		cancel()
+	}()
+	if _, err := conn.TryLock(interruptCtx, keyspacePath, "interrupted"); !topo.IsErrType(err, topo.NodeExists) {
+		require.Fail(t, "TryLock failed", err.Error())
+	}
```

Comment:
How does this test the cancellation? Shouldn't we be checking that the context being done is what caused the failure to lock?